src/vos/vos_layout.h

```diff
@@ -175,96 +173,96 @@ enum vos_dtx_record_types {
 	DTX_RT_EVT	= 3,
 };
 
-enum vos_dtx_record_flags {
-	/* We make some special handling for the punch of object/key.
-	 * The basic idea is that:
-	 *
-	 * Firstly, we insert a new record for the target (object/key)
-	 * to be punched with the given epoch.
-	 *
-	 * Then exchange the sub-tree(s) under the original record that
-	 * has DAOS_EPOCH_MAX for the target to be punched and the new
-	 * created record.
-	 *
-	 * And then remove the original record that has DAOS_EPOCH_MAX
-	 * for the target to be punched.
-	 */
-
-	/* The source of exchange the record that has the DAOS_EPOCH_MAX. */
-	DTX_RF_EXCHANGE_SRC	= 1,
-
-	/* The target of exchange the record that has the given epoch. */
-	DTX_RF_EXCHANGE_TGT	= 2,
+enum vos_dtx_entry_flags {
+	/* The DTX is the leader */
+	DTX_EF_LEADER			= (1 << 0),
+	/* The DTX entry is invalid. */
+	DTX_EF_INVALID			= (1 << 1),
 };
 
-/**
- * The agent of the record being modified via the DTX.
- */
+/** The agent of the record being modified via the DTX in both SCM and DRAM. */
 struct vos_dtx_record_df {
 	/** The DTX record type, see enum vos_dtx_record_types. */
-	uint32_t			tr_type;
-	/** The DTX record flags, see enum vos_dtx_record_flags. */
-	uint32_t			tr_flags;
-	/** The record in the related tree in SCM. */
-	umem_off_t			tr_record;
-	/** The next vos_dtx_record_df for the same DTX. */
-	umem_off_t			tr_next;
+	uint32_t			dr_type;
+	/** The 64-bits alignment. */
+	uint32_t			dr_padding;
+	/** The modified record in the related tree in SCM. */
+	umem_off_t			dr_record;
 };
 
-enum vos_dtx_entry_flags {
-	/* The DTX shares something with other DTX(s). */
-	DTX_EF_SHARES			= (1 << 0),
-	/* The DTX is the leader */
-	DTX_EF_LEADER			= (1 << 1),
-};
+#define DTX_INLINE_REC_CNT	4
```

Comment:
so this is 1 object + 1 dkey + 1 akey + 1 value right? we might bump it a little bit, e.g. to 6. so we are more likely to avoid allocation

Reply:
The case of update 1 object is only for the first time of creating the object.